### PR TITLE
Stratimikos, Piro, Panzer: Silence warnings

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/stratimikos @trilinos/panzer 

## Motivation
Silence some warnings (Shadowed variables, unused typedefs).

### CODE REVIEW

**Review by rppawlo — APPROVED** [review]
Thanks @cgcgcg !

File: packages/stratimikos/src/Stratimikos_LinearSolverBuilder_def.hpp [hunk 1]
```diff
@@ -88,7 +88,7 @@ LinearSolverBuilder<Scalar>::LinearSolverBuilder(
 
 
 template<class Scalar>
-LinearSolverBuilder<Scalar>::~LinearSolverBuilder<Scalar>()
+LinearSolverBuilder<Scalar>::~LinearSolverBuilder()
```
Comment (bartlettroscoe) [inline packages/stratimikos/src/Stratimikos_LinearSolverBuilder_def.hpp:91]: 👍

**Review by bartlettroscoe — APPROVED** [review]
Looks good.

Just curious, but what compilers/builds were emitting these warnings?  Why did the `-Werror` PR build not show these warnings?

**Comment by cgcgcg** [comment]
@bartlettroscoe I built with `-Werror=shadow -Werror=unused-variable -pedantic` in `CMAKE_CXX_FLAGS` and exceptions for shadow variables for some packages using gcc 14.2. I believe that the Werror builds use more exceptions:
https://github.com/trilinos/Trilinos/blob/bf922e75428ef7dd925811f0a532d05dd9cff808/packages/framework/ini-files/config-specs.ini#L1740-L1763
If someone from @trilinos/framework can confirm that that's the right spot to make modifications, I can remove the exceptions for Stratimikos and Piro. (I still see a Panzer warning coming in through ROL.)

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe I built with `-Werror=shadow -Werror=unused-variable -pedantic` in `CMAKE_CXX_FLAGS` and exceptions for shadow variables for some packages using gcc 14.2.

@cgcgcg, how did you think to build with these compiler flags?  Was a user complaining about these warnings?  Many of these warnings are not in header files so end users pulling in header files from Trilinos would not have seen those.  Just curious.

> If someone from @trilinos/framework can confirm that that's the right spot to make modifications, I can remove the exceptions for Stratimikos and Piro. (I still see a Panzer warning coming in through ROL.)

My guess is that there were a **lot** of shadow warnings in Trilinos and in order to get `-Werror` turned on, they disabled the shadow warnings so they could get all of the packages under `-Werror`.

**Comment by cgcgcg** [comment]
I added `-pedantic` to see #13299. And I caused #13358 and it was not caught by the AT.

**Comment by sebrowne** [comment]
@bartlettroscoe is correct about why those are all in there.  Remove them at will, the PR system should prevent issues from going in.  As to how your previous PR caused the issues, I'm not certain.  I can try to diagnose, but will likely take me some time.

Short answer is that as we turn on more warning flags (and stop allowing exceptions), this issue should rapidly become irrelevant.

**Pull request merged**